### ISSUE TITLE
NUMBER, INT and SCALAR field types

### ISSUE THREAD

**Issue by rtsisyk** [tarantool/tarantool#1556]
```
commit 13f106e2e5b67ef239de2b8b8156d3dc1355696e
Author: Vladislav Shpilevoy <v.shpilevoy@corp.mail.ru>
Date:   Thu Jun 16 14:57:53 2016 +0300

    Add new tree index types: NUMBER, INT and SCALAR
```

It seems that @Gerold103 and @alyapunov forgot to open a feature request for new field types. This feature is already implemented and @pgulutzan is already started to document it (#1551).

**Comment by rtsisyk** [comment]
There is a post-review fixes:

--------------------------------------------------------

1) `NULL` has special meaning and semantics in databases. We don't implement this semantics properly, therefore this feature is misleading for most users.

https://en.wikipedia.org/wiki/Null_%28SQL%29

Below are comments from @pgulutzan: 
When I have a SCALAR unique non-primary index, I cannot insert nil twice.
I regard this as odd. Is it deliberate?

When I have a SCALAR primary index, I can insert nil.
I regard this as odd. Is it deliberate?

When I have a SCALAR index, the order is: nil, booleans, numbers, strings.
This is wonderful, but is there a reason nil cannot also be in a NUMBER
index?

=>  Please remove `nil` support for now.

--------------------------------------------------------

2) `INT` field type for `HASH` index

I can create HASH index with 'INT', but there are no test cases for it.

```
tarantool> box.schema.space.create('fieldtypes')
tarantool> box.space.fieldtypes:create_index("primary", { type = 'HASH', parts = { 1, 'INT' }})
---
- unique: true
  parts:
  - type: INT
    fieldno: 1
  id: 0
  space_id: 513
  name: primary
  type: HASH
...
```

=> Please add tests cases for HASH with `INT` field type. 

--------------------------------------------------------

3) non-unique SCALAR, INT and NUMBER indexes

I can't find test cases.

=> Please add tests cases for `unique=false` with `SCALAR`, `INT`, `NUMBER` types

--------------------------------------------------------

4) `NUMBER` field type for `HASH` index

IEEE 754 floating point values are senseless for HASH index. Exact comparison of two floats is always bad idea. Moreover, it is very easy to create two floats with the same text representation but different binary value.

=> Please forbid `NUMBER` field type for `HASH`.

--------------------------------------------------------

5) `SCALAR` field type for `HASH` index

SCALAR has the same problem as NUMBER (see above).
There are three options:

* Keep it as it and add test cases
* Forbid SCALAR field type for HASH index
* Forbid float/double values in HASH index

Let's discuss these options.

--------------------------------------------------------

**Comment by kostja** [comment]
* Roman Tsisyk <notifications@github.com> [16/06/22 13:05]:

> 4) `NUMBER` field type for `HASH` index
> 
> IEEE 754 floating point values are senseless for HASH index. Exact comparison of two floats is always bad idea. Moreover, it is very easy to create two floats with the same text representation but different binary value.
> 
> => Please forbid `NUMBER` field type for `HASH`.
> 
> --------------------------------------------------------
> 
> 5) `SCALAR` field type for `HASH` index
> 
> SCALAR has the same problem as NUMBER (see above).
> There are three options:
> 
> * Keep it as it and add test cases
> * Forbid SCALAR field type for HASH index
> * Forbid float/double values in HASH index

I believe we should not ban it but state in the manual that the
behaviour is implementation defined.


-- 
Konstantin Osipov, Moscow, Russia, +7 903 626 22 32
http://tarantool.org - www.twitter.com/kostja_osipov

**Comment by rtsisyk** [comment]
Discussed with @alyapunov and @kostja 
1, 2, 3 OK 
4 and 5: => allow non-unique NUMBER and SCALAR indexes and document that the behavior is implementation defined.

**Comment by rtsisyk** [comment]
This problem is already resolved in 1.7.1. All required tests have been added.
